runtime/oti/VMHelpers.hpp

```diff
@@ -2068,9 +2070,61 @@ class VM_VMHelpers
 	static VMINLINE bool
 	isContinuationMounted(J9VMContinuation *continuation)
 	{
-		return (NULL != continuation->carrierThread);
+		return J9_ARE_ANY_BITS_SET(continuation->state, ~(uintptr_t)J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
+	}
+
+	static VMINLINE bool
+	isContinuationMountedOrConcurrentlyScanned(J9VMContinuation *continuation)
+	{
+		return isContinuationMounted(continuation) || isConcurrentlyScannedFromContinuationState(continuation->state);
 	}
+
+/*
+ *	If low tagging failed due to either
+ *
+ *   a carrier thread winning to mount, we don't need to do anything, since it will be compensated by pre/post mount actions
+ *   another GC thread winning to scan, again don't do anything, and let the winning thread do the work, instead
+ */
+	static VMINLINE bool
+	tryWinningConcurrentGCScan(J9VMContinuation *continuation)
+	{
+		return J9_GC_CONTINUATION_STATE_INITIAL == VM_AtomicSupport::lockCompareExchange(&continuation->state, J9_GC_CONTINUATION_STATE_INITIAL, J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
+	}
+
+	static VMINLINE void
+	exitConcurrentGCScan(J9VMContinuation *continuation)
+	{
+		/* clear CONCURRENTSCANNING flag */
+		uintptr_t oldContinuationState = VM_AtomicSupport::bitAnd(&continuation->state, ~(uintptr_t)J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
+		J9VMThread *carrierThread = getCarrierThreadFromContinuationState(oldContinuationState);
+		if (NULL != carrierThread) {
+			omrthread_monitor_enter(carrierThread->publicFlagsMutex);
+			/*  notify the waiting carrierThread that we just finished scanning, and it can proceed with mounting. */
+			omrthread_monitor_notify_all(carrierThread->publicFlagsMutex);
+			omrthread_monitor_exit(carrierThread->publicFlagsMutex);
+		}
+	}
+#endif /* JAVA_SPEC_VERSION >= 19 */
+
+	static VMINLINE UDATA
+	walkContinuationStackFramesWrapper(J9VMThread *vmThread, j9object_t continuationObject, J9StackWalkState *walkState, bool syncWithContinuationMounting)
+	{
+		UDATA rc = J9_STACKWALK_RC_NONE;
+#if JAVA_SPEC_VERSION >= 19
+		J9VMContinuation *continuation = J9VMJDKINTERNALVMCONTINUATION_VMREF(vmThread, continuationObject);
+		if (syncWithContinuationMounting && (NULL != continuation)) {
+			if (!tryWinningConcurrentGCScan(continuation)) {
+				/* If continuation is mounted or already being scanned by another GC thread, we do nothing */
+				return rc;
+			}
+		}
+		rc = vmThread->javaVM->internalVMFunctions->walkContinuationStackFrames(vmThread, continuation, walkState);
+		if (syncWithContinuationMounting && (NULL != continuation)) {
+			exitConcurrentGCScan(continuation);
+		}
 #endif /* JAVA_SPEC_VERSION >= 19 */
+		return rc;
+	}
 
 	/**
 	 * Check if we need to scan the java stack for the Continuation Object
```

Comment:
Add a comment: 
Used during main scan phase of GC (object graph traversal) or heap object iteration (in sliding compact). Not meant to be used during root scanning (neither strong roots nor weak roots)!